pkg/controller/controller_test.go

```diff
@@ -1960,13 +1962,91 @@ func TestProvision(t *testing.T) {
 			},
 			expectErr:          true,
 			expectState:        controller.ProvisioningNoChange,
-			expectSelectedNode: nodeFoo.Name,
+			expectNoProvision:  true,         // not owner yet
+			expectSelectedNode: nodeFoo.Name, // changed by ShouldProvision
+		},
+		"distributed immediate, allowed topologies okay": {
+			deploymentNode:   "foo",
+			immediateBinding: true,
+			volOpts: controller.ProvisionOptions{
+				StorageClass: &storagev1.StorageClass{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: fakeSCName,
+					},
+					ReclaimPolicy: &deletePolicy,
+					Parameters: map[string]string{
+						"fstype": "ext3",
+					},
+					VolumeBindingMode: &immediateBinding,
+					AllowedTopologies: []v1.TopologySelectorTerm{
+						{
+							MatchLabelExpressions: []v1.TopologySelectorLabelRequirement{
+								{
+									Key:    driverTopologyKey,
+									Values: []string{"foo"},
+								},
+							},
+						},
+					},
+				},
+				PVName: "test-name",
+				PVC:    createFakePVC(requestedBytes),
+			},
+			expectErr:          true,
+			expectState:        controller.ProvisioningNoChange,
+			expectNoProvision:  true,         // not owner yet
+			expectSelectedNode: nodeFoo.Name, // changed by ShouldProvision
 		},
+		"distributed immediate, allowed topologies not okay": {
+			deploymentNode:   "foo",
+			immediateBinding: true,
+			volOpts: controller.ProvisionOptions{
+				StorageClass: &storagev1.StorageClass{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: fakeSCName,
+					},
+					ReclaimPolicy: &deletePolicy,
+					Parameters: map[string]string{
+						"fstype": "ext3",
+					},
+					VolumeBindingMode: &immediateBinding,
+					AllowedTopologies: []v1.TopologySelectorTerm{
+						{
+							MatchLabelExpressions: []v1.TopologySelectorLabelRequirement{
+								{
+									Key:    driverTopologyKey,
+									Values: []string{"bar"},
+								},
+							},
+						},
+					},
+				},
+				PVName: "test-name",
+				PVC:    createFakePVC(requestedBytes),
+			},
+			expectErr:          true,
+			expectState:        controller.ProvisioningNoChange,
+			skipCreateVolume:   true,
+			expectNoProvision:  true, // not owner and will not change that either
```

Comment:
Prior to the fix, the test will fail at `expectNoProvision`, while `expectErr`, `expectState`, and `skipCreateVolume` will pass, right?

Reply:
It would have passed `expectNoProvision`, but for the wrong reason (not owner yet). Without the fix, it fails `expectSelectedNode` because the provisioner would set the current node to itself. The correct behavior is to not provision and not set the node.

Reply:
Gotcha. The behavior of selected node is a little non-obvious to me, and I had trouble making sense of what behavior should be expected from the test case parameters. Given enough time I should be able to figure it out but I couldn't quite grasp it just by scanning the test case.

Just a nit - not sure what could improve it but maybe a comment explaining the scenario could help?

Reply:
Maybe consider adding a comment that selecting a node == becoming owner

Reply:
Added a comment.